homeassistant/components/wiffi/sensor.py

```diff
@@ -0,0 +1,147 @@
+
+
+@callback
+def _create_entity(api, device, metric):
+    """Create platform specific entities."""
+    entities = []
+
+    if metric.is_number:
+        entities.append(NumberEntity(device, metric))
+    elif metric.is_string:
+        entities.append(StringEntity(device, metric))
+
+    stem = Path(__file__).stem  # stem = filename without py
+    api.async_add_entities[stem](entities)
+
+
+class NumberEntity(WiffiEntity):
+    """Entity for wiffi metrics which have a number value."""
+
+    def __init__(self, device, metric):
+        """Initialize the entity."""
+        super().__init__(device, metric)
+        self._device_class = determine_device_class(metric)
+        self._unit_of_measurement = convert_unit_of_measurement(
+            metric.unit_of_measurement
+        )
+        self._value = metric.value
+        self.reset_expiration_date()
+
+    @property
+    def device_class(self):
+        """Return the automatically determined device class."""
+        return self._device_class
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement of this entity."""
+        return self._unit_of_measurement
+
+    @property
+    def state(self):
+        """Return the value of the entity."""
+        return self._value
+
+    def _update_value(self, metric):
+        """Update the value of the entity.
+
+        Called if a new message has been received from the wiffi device.
+        """
+        self.reset_expiration_date()
+        self._unit_of_measurement = convert_unit_of_measurement(
+            metric.unit_of_measurement
+        )
+        self._value = metric.value
+        self.async_schedule_update_ha_state()
+
+
+class StringEntity(WiffiEntity):
+    """Entity for wiffi metrics which have a string value."""
+
+    def __init__(self, device, metric):
+        """Initialize the entity."""
+        super().__init__(device, metric)
+        self._value = metric.value
+        self.reset_expiration_date()
+
+    @property
+    def state(self):
+        """Return the value of the entity."""
+        return self._value
+
+    def _update_value(self, metric):
+        """Update the value of the entity.
+
+        Called if a new message has been received from the wiffi device.
+        """
+        self.reset_expiration_date()
+        self._value = metric.value
+        self.async_schedule_update_ha_state()
+
+
+def convert_unit_of_measurement(oum):
+    """Convert german wiffi texts to common HA texts."""
+    return UOM_MAP[oum] if oum in UOM_MAP else oum
+
+
+def determine_device_class(metric):
+    """Try to find best matching device class.
+
+    Currently only units of measurements are used for the detection and
+    therefore a dict with the uom as key could be used also. However, other
+    the detection may be improved in the future using other fields as well,
+    therefore a simple if/elsif chain is used here.
+    """
+    if metric.unit_of_measurement == WIFFI_UOM_TEMP_CELSIUS:
```

Comment:
Break out a module level constant dict that maps WIFFI UOM to device class. Then look up that dict here. We can then inline this operation in init method.